### PR TITLE
Run in integration-test phase

### LINKED ISSUE #15: CasperJS run in integration-test phase
Hi!

I'd like to run my casperJS tests in the `integration-test` phase, 'cause

 * I'll have to prepare a database with the correct data (in the `pre-integration-test` phase), 
 * I'll have to start a Jetty server with my application (in the `pre-integration-test` phase),
 * and I would like to stop the jetty server after the integration test run (in the `post-integration-test` phase)

If I config the plugin to run in the `integration` phase then if there are any failed test, the `post-integration-phase` won't run, and the Jetty server'll be stuck. Is there a way to run the tests in the `integration-test` phase, but fail in the `verify` phase, if there's any error? (Like maven failsafe plugin does)

Or do I have to write this myself?

Thanks,

### PR DESCRIPTION
**Pull request by vilmosnagy**
As discussed in #15.

I've created an example in the `it/integration-test`  directory, but I'm not sure, that I've configured the invoker plugin right.

Closes #15 .

### CODE REVIEW

**Comment by bguerin** [comment]
Thanks for your idea, your implementation and your share. I am ok with most of your modifications, but I you don't mind, I would have introduced an AbstractCasperJSRunnerMojo (mostly a rename from the actual CasperJSRunnerMojo), from which inherits :
- CasperJSRunnerTestMojo, mostly an empty class, just the @Mojo and the handling of the ignoreTestFailures parameter
- CasperJSRunnerIntegrationTestMojo, for your use case, which does not have the ignoreTestFailures parameter, never fails the build, and binds by default to the integration-test phase.
Doing this will simplifiy the the "client" POMs

File: src/it/casperjs-runner/integration-test/pom.xml [hunk 1]
```diff
@@ -0,0 +1,90 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>com.github.casperjs</groupId>
+  <artifactId>ignore-failures</artifactId>
```
Comment (bguerin) [inline src/it/casperjs-runner/integration-test/pom.xml]: should be changed to integration-test

File: src/it/casperjs-runner/integration-test/pom.xml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  <version>1.0</version>
+  <packaging>pom</packaging>
+
+  <name>CasperJS Runner :: ignore-failures</name>
```
Comment (bguerin) [inline src/it/casperjs-runner/integration-test/pom.xml]: should be changed to integration-test

File: src/main/java/com/github/casperjs/casperjsrunner/CasperJSVerifyMojo.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import org.apache.maven.plugins.annotations.Parameter;
+
+import java.io.*;
+
+/**
+ * Verifies that there was no test failure during the test Mojo.
+ *
+ * @author Vilmos Nagy
+ */
+@Mojo(name = "verify")
```
Comment (bguerin) [inline src/main/java/com/github/casperjs/casperjsrunner/CasperJSVerifyMojo.java]: could you add a "defaultPhase" to verify please ?

File: src/it/casperjs-runner/integration-test/pom.xml [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+          <execution>
+            <id>test</id>
+            <phase>integration-test</phase>
+            <goals>
+              <goal>test</goal>
+            </goals>
+          </execution>
+          <execution>
+            <id>verify</id>
+            <phase>verify</phase>
```
Comment (bguerin) [inline src/it/casperjs-runner/integration-test/pom.xml]: specifying the phase will be useless once you add a defaultPhase to the Mojo annotation on the Verify goal (see below)

File: src/main/java/com/github/casperjs/casperjsrunner/CasperJSRunnerMojo.java [hunk 5]
```diff
@@ -307,11 +317,34 @@ public void execute() throws MojoExecutionException, MojoFailureException {
...
         if (!ignoreTestFailures && globalResult.getFailures() > 0) {
             throw new MojoFailureException("There are " + globalResult.getFailures() + " tests failures");
         }
     }
 
+    private void writeFailedTestCount(Result globalResult) throws MojoExecutionException {
+        try {
+            tryToWriteFailedTestCount(globalResult);
+        } catch (IOException e) {
+            throw new MojoExecutionException("", e);
```
Comment (bguerin) [inline src/main/java/com/github/casperjs/casperjsrunner/CasperJSRunnerMojo.java]: could you please add a message to help understand the exception

File: src/main/java/com/github/casperjs/casperjsrunner/CasperJSRunnerMojo.java [hunk 6]
```diff
@@ -307,11 +317,34 @@ public void execute() throws MojoExecutionException, MojoFailureException {
...
+        }
+    }
+
+    private void tryToWriteFailedTestCount(Result globalResult) throws IOException {
+        BufferedWriter bufferedWriter = null;
+        try {
+            bufferedWriter = new BufferedWriter(new FileWriter(testFailureCountFile, false));
```
Comment (bguerin) [inline src/main/java/com/github/casperjs/casperjsrunner/CasperJSRunnerMojo.java]: you should take care of creating the parent directories of the destination file it they don't exist, I think this is what make CI fails

**Review by bguerin — CHANGES_REQUESTED** [review]

**Review by bguerin — APPROVED** [review]

**Comment by bguerin** [comment]
Thanks for all !

**Comment by vilmosnagy** [comment]
I've just finished, and wanted to resolve all the discussions above.
Thanks for the quick reply.

**Comment by bguerin** [comment]
Release 1.0.5 out with your PR, if you need it :)

**Pull request merged**